pandas/tseries/index.py

```diff
@@ -1257,6 +1257,12 @@ def get_loc(self, key, method=None):
                                           'when key is a time object')
             return self.indexer_at_time(key)
 
+        # check if it's a Period and the frequencies are the same - otherwise a monthly period would match for
+        # a daily timestamp at the beginning of the month. NB: 'B' and 'D' therefore won't match
+        if isinstance(key, com.ABCPeriod) and key.freq == self.freq:
```

Comment:
The additional check is if it's a `Period` - your examples returning `True` isn't a change.

Re the desired behavior - I don't have a confident view here - but I think those probably should return `True`. That's what `DatetimeIndex` is - just a series of stamps. I wouldn't have thought the `__contains__` method should have any knowledge about the frequency of the data - it's just asking if it's a member of the sequence. Anything else and people should be using `PeriodIndex`.

Where I could see this being an issue is where you have a `DatetimeIndex` that has dates at the _beginning_ of each month, and then you ask if a `Period` is in that. The `.to_timestamp()` method will convert to the end of month and `__contains__` will return `False`. I think that's an unlikely case given how difficult it seems to be to create such an index, but escalating to you regardless.

Am I missing something in the prior example?